extractor/src/main/java/org/schabi/newpipe/extractor/downloader/Response.java

```diff
@@ -2,29 +2,35 @@
 
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
-import java.util.Collections;
 import java.util.List;
 import java.util.Map;
+import java.util.SortedMap;
+import java.util.TreeMap;
 
 /**
  * A Data class used to hold the results from requests made by the Downloader implementation.
  */
 public class Response {
     private final int responseCode;
     private final String responseMessage;
-    private final Map<String, List<String>> responseHeaders;
+    private final SortedMap<String, List<String>> responseHeaders;
     private final String responseBody;
-
     private final String latestUrl;
 
     public Response(final int responseCode,
                     final String responseMessage,
-                    @Nullable final Map<String, List<String>> responseHeaders,
+                    @Nonnull final Map<String, List<String>> responseHeaders,
                     @Nullable final String responseBody,
                     @Nullable final String latestUrl) {
         this.responseCode = responseCode;
         this.responseMessage = responseMessage;
-        this.responseHeaders = responseHeaders == null ? Collections.emptyMap() : responseHeaders;
+
+        if (responseHeaders instanceof SortedMap) {
+            this.responseHeaders = (SortedMap<String, List<String>>) responseHeaders;
+        } else {
+            this.responseHeaders = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
+            this.responseHeaders.putAll(responseHeaders);
+        }
```

Comment:
Mmmh, this still means that all items are reprocessed when building a `Response` in case the headers were not already in a `SortedMap`. Also, how can you be sure that `responseHeaders` was built using `String.CASE_INSENSITIVE_ORDER`?

I would just build a new `TreeMap` every time.